server/main/templates/account/signup.html

```diff
@@ -5,21 +5,46 @@
 {% block head_title %}{% translate "Signup" %} &mdash; {{ block.super }}{% endblock %}
 
 {% block content %}
-    <form class="signup" id="signup_form login_form" method="post"
+    <form class="signup centered-form" id="signup_form" method="post"
             action="{% url 'account_signup' %}">
         {% csrf_token %}
-        <fieldset>
-            <legend>{% translate "Sign Up" %}</legend>
-            <p>
-                {% blocktranslate %}
-                Already have an account? Then please <a href="{{ login_url }}">sign in</a>.
-                {% endblocktranslate %}
-            </p>
-            {{ form.as_p }}
```

Comment:
Can you elaborate a little on the differences with hand-rolling the individual form elements vs using the standard output with `as_p`?

Reply:
Sure! In order to get the Bootstrap styles to properly apply, we need to add `.form-group` to the element that wraps each form field and its label. An additional part is that another class, `has-error`, needs to be added when the field is invalid in order for the red outline and error to display. I think your previous note about using a third party library might be the way to go long-term if a lot of this customization is needed, but since there aren't a ton of forms in the app, I'm not sure if that is worth it or not. Django's form helpers are really powerful, but controlling their HTML output to the degree required here is tricky.

An additional rub is that we want to display the password requirements to the user (it needs to be at least 6 characters long), but in the case of someone getting an error message on the field back from the backend, we are then presenting that information twice since we have both the initial instruction and the new validation error on the page. It's redundant, but ultimately not the end of the world to have that duplication.

If we don't mind the possibility of having that password info duplicated, it removes needing to special case that field and we can probably simplify this template down to something like this:

```html
{% for field in form.visible_fields %}
    <div class="form-group{% if field.errors %} has-error{% endif %}">
        {{ field.label_tag }}
        {{ field }}
        <div id="{{field.name}}_help" class="help-block">
            {{ f.errors }}
            {% if field.help_text %}
                <p>{{ field.help_text }}</p>
            {% endif %}
        </div>
    </div>
{% endif %}
```

What do you think?

Reply:
Thanks! I would have a small bias toward the simpler template, though the special-case does make sense now. I'll say move forward with this, then I will see if updating the templates used for form rendering to be Bootstrap-friendly can happen. It should be much easier in Django 4.0+, as all that moved to actual HTML templates, and not their weird factory function that each of the `as_p`, `as_ul` etc delegates to. There's still a few lagging packages keeping us from Django 4 the last time I checked.